includes/abstracts/abstract-wc-order.php

```diff
@@ -983,22 +983,47 @@ public function remove_coupon( $code ) {
 	 * @since  3.2.0
 	 */
 	protected function recalculate_coupons() {
-		$discounts = new WC_Discounts( $this );
-		$coupons   = $this->get_items( 'coupon' );
-
-		foreach ( $coupons as $coupon ) {
-			$coupon_object = new WC_Coupon( $coupon->get_code() );
-			$discounts->apply_coupon( $coupon_object );
-		}
-
-		$this->set_coupon_discount_amounts( $discounts );
-
 		// Reset line item totals.
 		foreach ( $this->get_items() as $item ) {
 			$item->set_total( $item->get_subtotal() );
 			$item->set_total_tax( $item->get_subtotal_tax() );
 		}
 
+		$discounts = new WC_Discounts( $this );
+
+		foreach ( $this->get_items( 'coupon' ) as $coupon_item ) {
+			$coupon_code   = $coupon_item->get_code();
+			$coupon_id     = wc_get_coupon_id_by_code( $coupon_code );
+			$coupon_object = false;
+
+			if ( $coupon_id ) {
+				$coupon_object = new WC_Coupon( $coupon_id );
+			} elseif ( $original_coupon_data = $coupon_item->get_meta( 'coupon_data', true ) ) {
+				$coupon_object = new WC_Coupon();
+				$coupon_object->set_props( $original_coupon_data );
```

Comment:
I will check it when testing, but do we have `is_virtual = true` within `$original_coupon_data` here? I assume we do.

Reply:
Fixed in 3cd1bedaded04fad5da22efd2e8a7c4e3c442454.